src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -381,8 +381,43 @@ private unsafe bool IsValueOfElementType(object value)
         // if this is an array of value classes and that value class has a default constructor
         // then this calls this default constructor on every element in the value class array.
         // otherwise this is a no-op.  Generally this method is called automatically by the compiler
+        public unsafe void Initialize()
+        {
+            MethodTable* pArrayMT = RuntimeHelpers.GetMethodTable(this);
+            var thElem = pArrayMT->GetArrayElementTypeHandle();
+            if (thElem.IsTypeDesc)
+            {
+                return;
+            }
+
+            var pElemMT = thElem.AsMethodTable();
+            if (!pElemMT->HasDefaultConstructor || !pElemMT->IsValueType)
+            {
+                return;
+            }
+
+            ref byte arrayRef = ref MemoryMarshal.GetArrayDataReference(this);
+            delegate*<ref byte, MethodTable*, void> ctorFtn = GetConstructorSlot(pElemMT);
+
+            nuint offset = 0;
+            for (var i = 0; i < Length; i++)
+            {
+                // Since GetConstructorSlot() is not idempotent and may have returned
```

Comment:
This is leaking a lot of low-level vm details into CoreLib. I think it would be better to use multicallable instantiated entrypoint for the constructor here. It would avoid the need to re-fetch the function pointer like this and it would avoid the need to pass the extra pMT argument to the constructor.

The problem is that fetching multicallable instantiated entrypoint is somewhat expensive and so it would need to be cached. Fortunately, we have a convenient place to cache this information on RuntimeType. It can look like this:

```
RuntimeType arrayType = GetType();

if (arrayType.GenericCache is not ArrayInitializeCache cache)
{
    cache = new ArrayInitializeCache(arrayType);
    arrayType.GenericCache = cache;
}

...

for (int i = 0; i < length; i++)
{
    ctorFtn(arrayRef);
    arrayRef = ref Unsafe.Add(arrayRef, elementSize);
}
```

This is also going to deal with the `EnsureInstanceActive` problem from the other comment. `EnsureInstanceActive` can be called while populating the cache.

Reply:
Ah, so we would be doing something similar to Activator. That makes sense. I suppose `Activiator.CreateInstance` is not so common for arrays, so it's OK if the cache has to be one or the other.

Reply:
Correct. Activator.CreateInstance does not work for arrays, so there is no concern about the two caches colliding.